flytekit/core/type_engine.py

```diff
@@ -486,14 +474,15 @@ def to_literal(self, ctx: FlyteContext, python_val: T, python_type: Type[T], exp
                 f"{type(python_val)} is not of type @dataclass, only Dataclasses are supported for "
                 f"user defined datatypes in Flytekit"
             )
-        if not self.is_serializable_class(type(python_val)):
-            raise TypeTransformerFailedError(
-                f"Dataclass {python_type} should be decorated with @dataclass_json or inherit DataClassJSONMixin to be "
-                f"serialized correctly"
-            )
+
         self._serialize_flyte_type(python_val, python_type)
 
-        json_str = python_val.to_json()  # type: ignore
+        if hasattr(python_val, "to_json"):
+            json_str = python_val.to_json()
```

Comment:
@wild-endeavor With this check, we are backward compatible. If one used `DataClassJsonMixin` or `@dataclass_json`, then `to_json` is called, which matches the original behavior. XREF: https://github.com/lidatong/dataclasses-json/blob/8512afc0a87053dbde52af0519c74198fa3bb873/dataclasses_json/api.py#L26

@Future-Outlier Can you include a comment here about how this preserves backward compatibility?

Reply:
Hi, @wild-endeavor and @thomasjpfan 

In flytekit-python, we use the function `to_json` to serialize dataclasses and `from_json` to deserialize them. Previously, we required users in older flytekit releases to add the`dataclass_json` decorator or inherit from the `DataclassJsonMixin` class.
This was because both provided the necessary methods to serialize (convert a dataclass to bytes) and deserialize (convert bytes back to a dataclass).

In this PR, the `mashumaro` module introduces two classes, `JSONEncoder` and `JSONDecoder`, for serializing and deserializing dataclasses.
These new classes eliminate the reliance on the `to_json` and `from_json` methods.

Initially, we used `if hasattr(python_val, "to_json"):` to check for the method's presence.
Therefore, introducing these changes will not cause any breaking changes.
This means that dataclasses inheriting from `DataclassJsonMixin` will continue to use `to_json` and `from_json` for serialization and deserialization when using this version of flytekit.


to_json REF: https://github.com/flyteorg/flytekit/blob/master/flytekit/core/type_engine.py#L498
from_json REF: https://github.com/flyteorg/flytekit/blob/master/flytekit/core/type_engine.py#L750

Reply:
To clarify, I wanted to have a short comment in the code that states how the "to_json" check, helps preserves backward compatibility.